docs/proposals/custom-metrics.md

```diff
@@ -0,0 +1,132 @@
+Custom metrics are needed by:
+* horizontal pod autoscaler, for autoscaling the number of pods based on them;
+* scheduler, for using them in a sophisticated scheduling algorithm.
+
+High level goals for our solution for version 1.2 are:
+* easy to use (it should be easy to export custom metric from user's application),
+* works for most application (it should be easy to configure monitoring for third-party applications),
+* performance & scalability (the largest supported cluster should be able to handle ~5 custom metrics per pod with reporting latency ~30 seconds).
+
+For version 1.2, we are not going to address the following issues (non-goals):
+* security of access to custom metrics,
+* general monitoring of application health by a user
+(out of the heapster scope, see [#665](https://github.com/kubernetes/heapster/issues/665)).
+
+## Design
+
+For the Kubernetes version 1.2, we plan to implement aggregation of pod custom metrics in Prometheus format by pull.
+
+Each pod, to expose custom metrics, will expose a set of Prometheus endpoints.
+(For version 1.2, we assume that custom metrics are not private information and they are accessible by everyone.
+In future, we may restrict it by making the endpoints accessible only by kubelet/cAdvisor).
+CAdvisor will collect metrics from such endpoints of all pods on each node by pulling, and expose them to Heapster.
+Heapster will:
+* collect custom metrics from all CAdvisors in the cluster, together with pulling system metrics
+(for version 1.2: we assume pooling period of ~30 seconds),
+* store them in a metrics backend (influxDB, Prometheus, Hawkular, GCM, …),
+* expose the latest snapshot of custom metrics for queries (by HPA/scheduler/…) using [model API](https://github.com/kubernetes/heapster/blob/master/docs/model.md).
+
+User can easily expose Prometheus metrics for her own application by using Prometheus [client](http://prometheus.io/docs/instrumenting/clientlibs/) library.
+To monitor third-party applications, Prometheus [exporters](http://prometheus.io/docs/instrumenting/exporters/) run as side-cars containers may be used.
+
+For version 1.2, to prevent a huge number of metrics negatively affect the system performance,
+the number of metrics that can be exposed by each pod will be limited to the configurable value (default: 5).
+
+In future versions (later than 1.2), we want to extend our solution by:
+* accepting pod metrics exposed in different formats than Prometheus
+(collecting of the different formats will need to be supported by cAdvisor),
+* support push metrics by exposing push API on heapster (e.g. in StatsD format) or on a local node collector
+(if heapster performance is insufficient),
+* support metrics not associated with an individual pod.
+
+
+## API
+
+For Kubernetes 1.2, defining pod Prometheus endpoints will be done using annotations.
+Later, when we are sure that our API is correct and stable, we will make it a part of `PodSpec`.
```

Comment:
I'm ok with the annotation approach for now. When we think about creating real API fields, I'd also like to think about what other things we might want in the "upward API".